### DISCUSSION TITLE
Easy configuration from mobile app

### DISCUSSION THREAD

**Discussion by shaman79** [arendst/Tasmota#19147] [General]
For my project I need to configure MQTT broker on Tasmota smart plug. As it will be done by end-users it needs to be easy and bulletproof process - just put the plug to the pairing mode, setup wifi and finalize the MQTT config.

MQTT broker address will be always the same, but I do not want users to enter it. The mobile app should do it automatically after the wifi setup.

What is the easiest way to achieve this? Should I upload .tapp file to the device and apply it or is there any other recommended way?

Thank you.

**Comment by Jason2866** [comment]
There is no official Tasmota APP, we don't know how and what your APP is doing for commissioning the device.
Since not telling what MCU you are using it is even more guess work.
Providing a fixed mqtt setup is easy. Just compile your firmware with this settings.

**Comment by shaman79** [comment]
Hello, we would like to use smart plug with Tasmota preinstalled, probably dlink. The idea is just to connect to the device, provide wifi auth and then setup mqtt.
We do not want to flash the fw as it is too demanding on the user side.
I am looking for a simple way to init Tasmota with our config from mobile app. Not sure if it is even possible, I am newbie with Tasmota.

**Reply by Jason2866** [reply]
You have to write the mobile APP, there is none. The APP has to search for AP Tasmota-xxxx connect to this AP and provide the needed settings. See https://tasmota.github.io/docs/Commands/
Since the DLink plug is esp8266 driven there is no Berry support -> no *.tapp files

Side note writing the APP will be "fun".  Since users do have iPhones and Android Devices. You have to write the APP for both platforms. Both APPs  needs certifying. Afaik for Apple certifying you are needing a paid Apple development Account.

**Reply by shaman79** [reply]
Cool, this was the direction I needed. So I will simply connect to Tasmota AP, use local ip and http commands to configure wifi and mqtt and thats it.

I already have experience with mobile app development. Android is ok, Apple is a bit of struggle with their old school processes, but we will make it I guess ;)

**Reply by sfromis** [reply]
FYI, @Staars seems to be working on using the Improv "open standard for connecting devices to Wi-Fi using Bluetooth LE" with Tasmota. Sounds relevant to you, but IDK what the current status is, or if/how it could work for your use case.

**Reply by Jason2866** [reply]
The local IP of the Tasmota AP is always `192.168.4.1`

**Reply by Jason2866** [reply]
@sfromis Since esp8266 is used, no BLE. The BLE Improv from Staars is already working. But the Improv standard does handle only wifi settings. So even a device with esp32x as MCU is used, it would not be a solution for this use case.
EDIT: The Berry part could be changed to handle mqtt stuff too. This would need changes in the (not released nor available) APP code too.

**Comment by Staars** [comment]
The Improv standard needs BLE, except for serial connection, which we already use in the web flasher.

IMHO it is really hard to create a mobile app that scans for for AP's and then connects to it, without constantly fighting the underlying OS (Android and iOS) - not saying it is impossible.
This is where BLE shines and i.e. the Matter guys did probably choose it for wireless commissioning for a good reason.

All I can say for now is, that it is feasible to use Improv-wifi via BLE for Tasmota and that I can recommend Flutter to accomplish this. Without any platform agnostic code my test app runs on Android, iOS and macOS in the same way. Extending the Improv standard for a personal project would be easy (like setting up MQTT).

**Reply by shaman79** [reply]
Do you have possibly tip for smart plug with Tasmota preinstalled and BLE support? Ideally something you use for testing. Also, would you be willing to share the code? I was working with BLE before, but it could save some time with prototyping.

**Reply by Staars** [reply]
I am not aware of any smart plug with preinstalled Tasmota with full BLE support.
You can just flash any of the mi32-homebridge-variants from Tasmota-specials, which will launch the Improv driver after fresh install or `reset 1`.

The prototype app for the mobile is here: https://github.com/Staars/improv_wifi_ble
This is my rolling testing ground and not a polished consumer product. UI and error handling is not complete.

Keep in mind, that you open a new door to your devices, which will raise new security concerns. That’s why I only activated the BLE server, when the device is „open“ anyway. In a real world scenario a mandatory user interaction like a button press should be used (part of the Improv protocol).
